controller/backing_image_controller.go

```diff
@@ -291,9 +304,117 @@ func (bic *BackingImageController) syncBackingImage(key string) (err error) {
+	for diskUUID := range bi.Spec.DiskFileSpecMap {
+		fileStatus, exists := bi.Status.DiskFileStatusMap[diskUUID]
+		if !exists || (fileStatus.State != longhorn.BackingImageStateFailed &&
+			fileStatus.State != longhorn.BackingImageStateFailedAndCleanUp &&
+			fileStatus.State != longhorn.BackingImageStateUnknown) {
+			// Non-existing file in status could due to not being synced from backing image manager yet.
+			// Consider it as newly created copy and count it as non-failed copies.
+			// So we don't create extra copy when handling copies evictions.
+			nonFailedCopies += 1
+		}
+	}
+
+	// Need to count the evicted copies in the nonFailedCopies then handle it in handleBackingImageCopiesEvictions
+	// so we can distinguish the case of "0 healthy copy" and "there is 1 copy but being evicted".
+	if nonFailedCopies == 0 {
+		return nil
+	} else {
+		nonEvictingCount := nonFailedCopies
+		for _, fileSpec := range bi.Spec.DiskFileSpecMap {
+			if fileSpec.EvictionRequested {
+				nonEvictingCount--
+			}
+		}
+
+		if nonEvictingCount < bi.Spec.MinNumberOfCopies {
+			readyNode, readyDiskName, err := bic.ds.GetReadyNodeDiskForBackingImage(bi)
+			logrus.Infof("replicate the copy to node: %v, disk: %v", readyNode, readyDiskName)
+			if err != nil {
+				logrus.WithError(err).Warnf("failed to create the backing image copy")
+				return nil
+			}
+			// BackingImageManager will then sync the BackingImage to the disk
+			bi.Spec.DiskFileSpecMap[readyNode.Status.DiskStatus[readyDiskName].DiskUUID] = &longhorn.BackingImageDiskFileSpec{}
+		}
+	}
+
+	return nil
+}
+
+func (bic *BackingImageController) cleanupEvictionRequestedBackingImageCopies(bi *longhorn.BackingImage) {
+	log := getLoggerForBackingImage(bic.logger, bi)
+
+	// If there is no non-evicting healthy backing image copy,
+	// Longhorn should retain one evicting healthy backing image copy for replenishing.
+	hasNonEvictingHealthyBackingImageCopy := false
+	evictingHealthyBackingImageCopyDiskUUID := ""
+	for diskUUID, fileSpec := range bi.Spec.DiskFileSpecMap {
+		fileStatus := bi.Status.DiskFileStatusMap[diskUUID]
+		if fileStatus == nil { // it is newly added, consider it as non healthy
+			continue
+		} else {
+			if fileStatus.State != longhorn.BackingImageStateReady {
+				continue
+			}
+		}
+
+		if !fileSpec.EvictionRequested {
+			hasNonEvictingHealthyBackingImageCopy = true
+			break
+		}
+		evictingHealthyBackingImageCopyDiskUUID = diskUUID
+	}
+
+	for diskUUID, fileSpec := range bi.Spec.DiskFileSpecMap {
+		if !fileSpec.EvictionRequested {
+			continue
+		}
+		if !hasNonEvictingHealthyBackingImageCopy && diskUUID == evictingHealthyBackingImageCopyDiskUUID {
+			msg := fmt.Sprintf("Failed to evict backing image copy on disk %v for now since there is no other healthy backing image copy", diskUUID)
+			log.Warn(msg)
+			bic.eventRecorder.Event(bi, corev1.EventTypeNormal, constant.EventReasonFailedDeleting, msg)
+			continue
+		}
+		delete(bi.Spec.DiskFileSpecMap, diskUUID)
```

Comment:
NIT: Add a comment to explain this special spec update

Reply:
added